### ISSUE TITLE
send_sms die-ing change

### ISSUE THREAD

**Issue by jwittkoski** [mbeijen/SMS-Send-Twilio#4]
I was confused about commit 00458bb6d578bff53faa09049e02eadd1880cae7 and die-ing if send_sms fails.

The POD docs say this about `new`:
> Returns a new SMS::Send object, or dies on error.

But regarding `send_sms` they say:
> The send_sms returns true if the message was sent, or the driver is fire-and-forget and unable to determine success, or false if the message was not sent.

Also, the code for `SMS::Send::send_sms` has:
```perl
        # Merge params and hand off
        my $rv = $self->_OBJECT_->send_sms(
                text => $text,
                to   => $to,
                $self->_PRIVATE(@_),
        );
 
        # Verify we get some sort of result
        unless ( defined $rv ) {
                Carp::croak("Driver did not return a result");
        }
 
        return $rv;
```

So the code that is calling the Twilio driver only croaks if an undef is returned. It's expecting a true/false from the driver which is returned to the caller.

I can understand die/croak-ing if something goes wrong in `new` but die/croak-ing in `send_sms` doesn't seem to match what SMS::Send expects. In the situation where the driver is working and configured but there is a problem with sending to Twilio, I would expect a true/false result.

**Comment by jwittkoski** [comment]
I do agree there should be some way to return more details about the driver/provider failure instead of just returning `false` or `croak`-ing. Seems like SMS::Send should have an `error` or `errstr` method for drivers to pass back failure information that the caller could access.

**Comment by jwittkoski** [comment]
Ok, I found the docs you referenced (SMS::Send::Driver):

> Should return true for either success or fire-and-forget with unknown result, defined-but-false ('' or 0) for a failed message send, or die on a fatal error.

This wording implies that failure to send or deliver a message is not a fatal error.

The other general concern is the change in *behavior* - current users of SMS::Send::Twilio who were not expecting to have to catch an exception when a send failed. Previously they just checked the return value, now their app may die unexpectedly.

**Comment by mbeijen** [comment]
I agree with you that it would be very nice if SMS::Send had some kind of generic error message handling and reporting from the driver.

I changed the behaviour of my Twilio module back to before and now it reports:

`Driver did not return a result at t/twilio.t line 34.`

when I try to send without network or with an incorrect authtoken.
This IS the original behaviour - the twilio driver returns undef which causes the send() call dies because SMS::Send. 

So all in all now it's the Twilio driver that dies and at least puts a helpful error message in `$@`.
Did you run into trouble with my change? If so, what trouble? In that case I might be tempted to revert it.

**Comment by jwittkoski** [comment]
I have some (not yet production) changes to a project using SMS::Send::Twilio and when I upgraded to 0.18 some of my tests broke.

Prior to 0.18 a failure in SMS::Send::Twilio::send_sms returned 0 for many real world cases:

* Unauthorized (wrong AccountSid or AuthToken)
* Invalid phone number (from)
* Invalid phone number (to)
* Text length is > 160 characters

Now these cases croak instead.

The conditions above are valid JSON responses from the API so my reading of SMS::Send::Driver is they shouldn't be considered a fatal error (but I know not everyone will agree on this).

I understand why the you made the change though - the inability to get useful failure information with the current SMS::Send design is frustrating.

It's not difficult for me to adjust to the new behavior as this is new code (and the ability to get failure info is helpful), but it is a change in behavior that could trip up other users.
